File: pdv-python/pdv/default_handlers.py
"""
pdv.default_handlers — Built-in double-click plot handlers for common
scientific Python types.

Registers ``@pdv.handle(...)`` entries so that double-clicking a tree
node whose value is an ``np.ndarray``, ``pd.Series``, ``pd.DataFrame``,
or ``xr.DataArray`` opens a matplotlib window with a sensible default
plot.

Registration is **lazy**: :func:`register_defaults` only registers
handlers for libraries that are already in ``sys.modules``, and the
handler-registry lookups (:func:`pdv.modules.has_handler_for`,
``dispatch_handler``) call it on every lookup. Importing numpy, pandas,
and xarray eagerly at kernel bootstrap cost real startup latency and
undercut the never-import-xarray design in ``serialization.py`` — and a
tree value can only *be* one of these types if its library is already
imported, so the sys.modules gate loses nothing.

Users can override any default by registering their own handler for the
same type in a module. Defaults never overwrite an existing
registration, so a user handler wins regardless of registration order.

Behavior summary
----------------
- ``np.ndarray``: 1D → ``ax.plot``; 2D → ``ax.imshow`` + colorbar;
  0D or >2D → print a notice to stdout, no plot.
- ``pd.Series``: ``series.plot()``.
- ``pd.DataFrame``: ``df.plot()`` (pandas plots all numeric columns vs index).
- ``xr.DataArray``: ``da.plot()`` (xarray dispatches 1D → line, 2D →
  pcolormesh, >2D → histogram based on dimensionality).

The path of the tree node is used as the figure title in every case so
the user can tell which node a window belongs to.

If a value cannot actually be plotted (e.g. a non-numeric ``pd.Series``
or an object-dtype array), the handler closes the half-built figure and
prints a ``[PDV]`` notice instead of raising — a raised exception would
surface to the renderer as an opaque ``internal.error`` comm response.
"""

from __future__ import annotations

from typing import Any, Callable


# Libraries whose defaults have been registered since the last reset.
# Latch so the per-lookup register_defaults() call is a cheap set check.
_lazy_done: set[str] = set()


def register_defaults() -> None:
    """Register built-in plot handlers for every library already imported.

    Called lazily from the handler-registry lookups
    (:func:`pdv.modules.has_handler_for` / ``dispatch_handler``) rather
    than eagerly at kernel bootstrap — see the module docstring for why.
    Checking ``sys.modules`` never triggers an import.

    Never overwrites an existing registration (user handlers win). Safe
    to call any number of times; each library is processed once until
    :func:`pdv.modules.clear_handlers` resets the latch.
    """
    import sys  # noqa: PLC0415

    for lib_name, registrar in (
        ("numpy", _register_numpy),
        ("pandas", _register_pandas),
        ("xarray", _register_xarray),
        ("h5py", _register_h5py),
    ):
        if lib_name in _lazy_done or lib_name not in sys.modules:
            continue
        _lazy_done.add(lib_name)
        registrar(_default_handle)


def _reset_lazy_state() -> None:
    """Forget which libraries' defaults were registered.

    Called by :func:`pdv.modules.clear_handlers` so a registry reset
    (tests, primarily) gets defaults re-registered on the next lookup.
    """
    _lazy_done.clear()


def _default_handle(cls: type) -> Callable:
    """``pdv.handle`` variant for builtin defaults.

    Skips registration when *cls* already has a handler, so a
    user-registered handler for the same type always wins over the
    builtin default — regardless of which registered first.
    """

    def decorator(func: Callable) -> Callable:
        from pdv.modules import _handler_registry, handle  # noqa: PLC0415

        if cls in _handler_registry:
            return func
        return handle(cls)(func)

    return decorator


def _plot_or_notice(
    path: str, draw: Callable[[Any, Any], None], *, ncols: int = 1
) -> None:
    """Create a fresh figure, run ``draw(fig, ax)``, title it, and show it.

    On any failure — whether in ``draw`` or in the final ``plt.show()`` —
    close the figure and print a ``[PDV]`` notice instead of raising:
    handler exceptions would otherwise propagate through ``dispatch_handler``
    and reach the renderer as an opaque ``internal.error``.

    Exactly one figure is created per call, so both inline paths — the
    ``matplotlib-inline`` backend enabled by :mod:`pdv.mpl_config` and its
    legacy ``plt.show()`` shim (which captures ``plt.gcf()``) — always
    resolve to this figure. The success path leaves the figure for
    ``plt.show()`` to own; any failure path closes it.

    Parameters
    ----------
    path : str
        Dot-separated tree path of the node; used as the axes title
        (figure suptitle when ``ncols > 1``).
    draw : Callable[[Figure, Axes], None]
        Callback that draws onto the supplied figure and axes. With
        ``ncols > 1`` the second argument is the ndarray of axes that
        ``plt.subplots(1, ncols)`` returns.
    ncols : int
        Number of side-by-side axes to create (default 1, the historical
        single-axes contract).
    """
    import matplotlib.pyplot as plt  # noqa: PLC0415

    fig, axes = plt.subplots(1, ncols)
    try:
        draw(fig, axes)
        if ncols == 1:
            axes.set_title(path)
        else:
            fig.suptitle(path)
    except Exception as exc:  # noqa: BLE001
        plt.close(fig)
        print(f"[PDV] Cannot plot {path!r}: {exc}")
        return
    try:
        plt.show()
    except Exception as exc:  # noqa: BLE001
        plt.close(fig)
        print(f"[PDV] Could not display {path!r}: {exc}")


def _plot_array_like(obj: Any, path: str, noun: str = "ndarray") -> None:
    """Plot a 1-D array as a line or a 2-D array as an image.

    Shared drawing logic for the ndarray and h5py.Dataset handlers.
    Complex arrays split into real and imaginary parts: 1-D plots Re and
    Im as two labeled lines on one axes; 2-D shows side-by-side Re/Im
    images, each with its own colorbar (the two parts routinely span
    different ranges, so a shared scale would flatten one of them).
    Higher-dimensional arrays print a ``[PDV]`` notice instead.

    Parameters
    ----------
    obj : array-like
        Object with ``ndim``/``shape``/``dtype`` and array semantics.
    path : str
        Dot-separated tree path, used as the plot title.
    noun : str
        Type noun for the can't-plot notice (e.g. ``'h5py dataset'``).
    """
    import numpy as np  # noqa: PLC0415

    if np.iscomplexobj(obj):
        arr = np.asarray(obj)
        if arr.ndim == 1:

            def _draw_lines(fig: Any, ax: Any) -> None:
                ax.plot(arr.real, label="Re")
                ax.plot(arr.imag, label="Im")
                ax.legend()

            _plot_or_notice(path, _draw_lines)
        elif arr.ndim == 2:

            def _draw_panels(fig: Any, axes: Any) -> None:
                ax_re, ax_im = axes
                im_re = ax_re.imshow(arr.real)
                ax_re.set_title("Re")
                fig.colorbar(im_re, ax=ax_re)
                im_im = ax_im.imshow(arr.imag)
                ax_im.set_title("Im")
                fig.colorbar(im_im, ax=ax_im)

            _plot_or_notice(path, _draw_panels, ncols=2)
        else:
            print(
                f"[PDV] Cannot plot {arr.ndim}-D complex {noun} "
                f"(shape={tuple(arr.shape)}); default handler supports "
                f"1D and 2D only."
            )
        return
    if obj.ndim == 1:
        _plot_or_notice(path, lambda fig, ax: ax.plot(obj))
    elif obj.ndim == 2:

        def _draw(fig: Any, ax: Any) -> None:
            im = ax.imshow(obj)
            fig.colorbar(im, ax=ax)

        _plot_or_notice(path, _draw)
    else:
        print(
            f"[PDV] Cannot plot {obj.ndim}-D {noun} (shape={tuple(obj.shape)}); "
            f"default handler supports 1D and 2D only."
        )


def _register_numpy(handle: Any) -> None:
    try:
        import numpy as np  # noqa: PLC0415
    except ImportError:
        return

    @handle(np.ndarray)
    def _on_ndarray(obj: Any, path: str, pdv_tree: Any) -> None:
        _plot_array_like(obj, path)


def _register_pandas(handle: Any) -> None:
    try:
        import pandas as pd  # noqa: PLC0415
    except ImportError:
        return

    @handle(pd.Series)
    def _on_series(obj: Any, path: str, pdv_tree: Any) -> None:
        _plot_or_notice(path, lambda fig, ax: obj.plot(ax=ax))

    @handle(pd.DataFrame)
    def _on_dataframe(obj: Any, path: str, pdv_tree: Any) -> None:
        _plot_or_notice(path, lambda fig, ax: obj.plot(ax=ax))


def _register_xarray(handle: Any) -> None:
    try:
        import xarray as xr  # noqa: PLC0415
    except ImportError:
        return

    @handle(xr.DataArray)
    def _on_dataarray(obj: Any, path: str, pdv_tree: Any) -> None:
        # xarray's DataArray.plot() dispatches by dimensionality:
        # 1D → line, 2D → pcolormesh, anything else → histogram. All three
        # draw onto the axes passed via ``ax=``.
        _plot_or_notice(path, lambda fig, ax: obj.plot(ax=ax))


# Materializing an h5py dataset for plotting reads it fully into memory —
# cap it so double-clicking a huge dataset can't freeze the kernel. 1-D/2-D
# arrays beyond this are for slicing in code, not a default line/image plot.
_H5PY_PLOT_MAX_BYTES = 100 * 1024 * 1024


def _register_h5py(handle: Any) -> None:
    try:
        import h5py  # noqa: PLC0415
        import numpy as np  # noqa: PLC0415
    except ImportError:
        return

    @handle(h5py.Dataset)
    def _on_h5py_dataset(obj: Any, path: str, pdv_tree: Any) -> None:
        # Only numeric/bool datasets have a default plot — decide from the
        # dtype BEFORE any read. Also load-bearing for the cap: vlen/object
        # dtypes report a bogus 8-byte itemsize (the pointer, not the
        # payload), so the size estimate below would wildly undercount a
        # string dataset and materialize it just to fail plotting (review).
        if not (
            np.issubdtype(obj.dtype, np.number)
            or np.issubdtype(obj.dtype, np.bool_)
        ):
            print(
                f"[PDV] No default plot for h5py dataset at {path!r} "
                f"(dtype {obj.dtype})"
            )
            return
        nbytes = obj.size * obj.dtype.itemsize
        if nbytes > _H5PY_PLOT_MAX_BYTES:
            print(
                f"[PDV] Cannot plot {path!r}: dataset is "
                f"{nbytes / 1e6:.0f} MB (cap "
                f"{_H5PY_PLOT_MAX_BYTES / 1e6:.0f} MB). Slice it in code, "
                f"e.g. pdv_tree[{path!r}][::10]."
            )
            return
        _plot_array_like(np.asarray(obj), path, noun="h5py dataset")

File: pdv-julia/src/default_handlers.jl
# default_handlers.jl — Built-in double-click handlers for common types.
#
# Port of pdv/default_handlers.py. Double-clicking a tree node whose value is
# a numeric or complex Vector/Matrix, a Makie Figure, a DataFrame, or an
# HDF5 dataset (virtual child of a PDVHdf5 node) does something sensible
# without the user writing a module.
#
# Registration is **lazy**, mirroring Python's sys.modules gate:
# `register_default_handlers!()` runs from every registry lookup
# (`has_handler_for` / `dispatch_handler`) but short-circuits on a
# loaded-module count latch, so the per-lookup cost is one integer compare.
# A tree value can only *be* a Makie Figure or DataFrame if its library is
# already loaded, so the gate loses nothing — and PDVKernel never has to
# depend on (or import) any plotting package.
#
# Two registration mechanisms, chosen per type:
# - Base types (numeric AbstractVector/AbstractMatrix) get ordinary
#   `pdv_handle` methods at package load; multiple dispatch means a user's
#   more-specific method automatically wins. The body runtime-gates on a
#   Makie backend and prints a `[PDV]` notice when none is loaded.
# - Foreign types (Makie.Figure, FigureAxisPlot, DataFrame) are registered
#   through `register_handler` once their library appears in
#   `Base.loaded_modules` — the types don't exist at package load time.
#   Defaults never overwrite an existing registration (user handlers win).
#
# Makie figures additionally get a lazy `pdv_digest` method that hashes the
# rendered pixels (`colorbuffer`). Unlike a structural walk of the live
# object graph (display state, weak refs), the rendered image is a
# content-faithful digest that is stable across a save/load round trip, so
# reopened figure-bearing projects don't show a spurious checksum-mismatch
# marker.

# Loaded-module count at the last lazy scan. Any package load changes the
# count, which is the only event that can make a new default registrable.
const _defaults_seen_module_count = Ref(0)

# Libraries whose defaults have been registered (latch, like Python's
# _lazy_done). Reset by clear_handlers! via _reset_default_handlers!.
const _defaults_done = Set{Symbol}()

"""
    register_default_handlers!()

Register built-in handlers for every supported library that is already
loaded. Called from the handler-registry lookups rather than eagerly at
bootstrap; cheap (one integer compare) unless a new module was loaded since
the previous call. Never overwrites an existing registration.
"""
function register_default_handlers!()
    n = length(Base.loaded_modules)
    n == _defaults_seen_module_count[] && return nothing
    _defaults_seen_module_count[] = n

    if !(:Makie in _defaults_done)
        mk = loaded_module(:Makie)
        if mk !== nothing
            push!(_defaults_done, :Makie)
            _register_makie_defaults!(mk)
        end
    end
    if !(:DataFrames in _defaults_done)
        df = loaded_module(:DataFrames)
        if df !== nothing
            push!(_defaults_done, :DataFrames)
            _register_dataframes_defaults!(df)
        end
    end
    if !(:HDF5 in _defaults_done)
        h5 = loaded_module(:HDF5)
        if h5 !== nothing
            push!(_defaults_done, :HDF5)
            _register_hdf5_defaults!(h5)
        end
    end
    nothing
end

"""Forget the lazy-registration state (tests; called by clear_handlers!)."""
function _reset_default_handlers!()
    empty!(_defaults_done)
    _defaults_seen_module_count[] = 0
    nothing
end

# Register `func` for `T` only when nothing is registered yet — a
# user-registered handler for the same type always wins over a default.
function _register_default(func::Function, T::Type)
    haskey(_HANDLER_REGISTRY, T) && return nothing
    _HANDLER_REGISTRY[T] = HandlerEntry(T, func, fully_qualified_type_name(T))
    nothing
end

function _register_makie_defaults!(mk::Module)
    # Resilient display: a reloaded figure can exist with Makie loaded but no
    # backend activated (nothing to render with) — tell the user what to do
    # instead of surfacing an opaque handler error.
    show_fig = (obj, path, tree) -> try
        Base.invokelatest(display, obj)
    catch err
        println("[PDV] Cannot display '$path': $(sprint(showerror, err))\n" *
                "[PDV] Load a Makie backend (e.g. `using CairoMakie`) and try again.")
    end
    isdefined(mk, :Figure) && _register_default(show_fig, getproperty(mk, :Figure))
    isdefined(mk, :FigureAxisPlot) &&
        _register_default(show_fig, getproperty(mk, :FigureAxisPlot))

    # Rendered-pixels digest for figures (see module docstring). Evaluated
    # lazily because the Figure type doesn't exist at package load time.
    if isdefined(mk, :Figure)
        FigT = getproperty(mk, :Figure)
        if !hasmethod(pdv_digest, Tuple{FigT})
            Core.eval(@__MODULE__,
                      :(pdv_digest(fig::$FigT) = _makie_figure_digest(fig)))
        end
    end
    nothing
end

function _register_dataframes_defaults!(df::Module)
    isdefined(df, :DataFrame) || return nothing
    _register_default((obj, path, tree) -> Base.invokelatest(display, obj),
                      getproperty(df, :DataFrame))
    nothing
end

# Double-clicking an HDF5 dataset (a virtual child of a PDVHdf5 node) never
# silently materializes more than this many bytes into memory. A Ref so
# tests can lower it without a 100 MB fixture.
const _HDF5_PLOT_MAX_BYTES = Ref(100_000_000)

function _register_hdf5_defaults!(h5::Module)
    isdefined(h5, :Dataset) || return nothing
    _register_default(_hdf5_dataset_handler, getproperty(h5, :Dataset))
    nothing
end

# Default handler for HDF5.Dataset: materialize (under the size cap) and
# reuse the numeric/complex array plot methods (same contract as Python's
# h5py.Dataset handler).
function _hdf5_dataset_handler(obj, path, tree)
    nbytes = try
        n = Base.invokelatest(length, obj)
        T = Base.invokelatest(eltype, obj)
        n * max(sizeof(T), 1)
    catch
        0
    end
    if nbytes > _HDF5_PLOT_MAX_BYTES[]
        mb = round(nbytes / 1_000_000; digits=1)
        cap_mb = max(_HDF5_PLOT_MAX_BYTES[] ÷ 1_000_000, 1)
        println("[PDV] Dataset at '$path' is $(mb) MB — larger than the " *
                "$(cap_mb) MB default-plot cap. Read a slice instead, e.g. " *
                "pdv_tree[\"$path\"][1:1000].")
        return nothing
    end
    data = try
        Base.invokelatest(read, obj)
    catch err
        println("[PDV] Cannot read dataset at '$path': $(sprint(showerror, err))")
        return nothing
    end
    if data isa AbstractArray && (eltype(data) <: Real || eltype(data) <: Complex)
        Base.invokelatest(pdv_handle, data, String(path), tree)
    else
        println("[PDV] No default plot for dataset at '$path' " *
                "(element type $(typeof(data)))")
    end
    nothing
end

"""
    _makie_figure_digest(fig) -> Vector{UInt8}

Digest payload for a Makie figure: the rendered image bytes prefixed with
the pixel dimensions. Rendering is deterministic for identical figure
content, so the digest survives a save/load round trip (unlike the live
object graph). Errors propagate — the checksum walk treats a failed
`pdv_digest` as "no digest" and falls back to its own scheme.
"""
function _makie_figure_digest(fig)::Vector{UInt8}
    mk = loaded_module(:Makie)
    mk === nothing && error("Makie is not loaded")
    colorbuffer = getproperty(mk, :colorbuffer)
    img = try
        Base.invokelatest(colorbuffer, fig)
    catch err
        # A freshly-reloaded project has Makie (the figure's own package)
        # loaded but no backend activated — rendering needs one. Pull in
        # CairoMakie (the standard in-memory renderer) and retry once, so
        # post-load checksums keep matching their save-time values. If it
        # isn't installed the original error propagates and the checksum
        # walk falls back to its own scheme.
        loaded_module(:CairoMakie) === nothing || rethrow()
        Base.require(Main, :CairoMakie)
        Base.invokelatest(colorbuffer, fig)
    end
    header = collect(reinterpret(UInt8, [htol(UInt64(size(img, 1))),
                                         htol(UInt64(size(img, 2)))]))
    return vcat(header, collect(reinterpret(UInt8, vec(img))))
end

# ---------------------------------------------------------------------------
# Numeric-array defaults (Base types: plain pdv_handle methods)
# ---------------------------------------------------------------------------

# Tests disable the backend auto-load so suites stay deterministic (and
# never drag a multi-second CairoMakie load into a unit test) regardless of
# whether CairoMakie is resolvable from the test environment stack.
const _MAKIE_AUTOLOAD_ENABLED = Ref(true)

# When no Makie backend is loaded but CairoMakie is installed in the active
# environment, load it and return the Makie module — the Julia analog of
# Python's default handlers importing matplotlib on first plot (matplotlib
# is a hard dep there; CairoMakie is optional here but prefilled into new
# projects, so first double-click should plot, not lecture). Same precedent
# as _makie_figure_digest's CairoMakie auto-require. Returns nothing when
# unavailable or the load fails.
function _try_autoload_makie_backend(path::String)
    _MAKIE_AUTOLOAD_ENABLED[] || return nothing
    Base.identify_package("CairoMakie") === nothing && return nothing
    println("[PDV] Loading CairoMakie to plot '$path' (first plot in this session)…")
    try
        Base.require(Main, :CairoMakie)
    catch err
        println("[PDV] Failed to load CairoMakie: $(sprint(showerror, err))")
        return nothing
    end
    return loaded_module(:Makie)
end

# Run `draw(makie_module)` and display what it returns, or print a `[PDV]`
# notice — a raised exception would reach the renderer as an opaque
# internal.error (same contract as Python's _plot_or_notice). Function-first
# so call sites can use do-block syntax.
function _plot_with_makie(draw::Function, path::String)
    mk = loaded_module(:Makie)
    mk === nothing && (mk = _try_autoload_makie_backend(path))
    if mk === nothing
        println("[PDV] Cannot plot '$path': no Makie backend is loaded. " *
                "Run e.g. `using CairoMakie` first.")
        return nothing
    end
    try
        result = Base.invokelatest(draw, mk)
        Base.invokelatest(display, result)
    catch err
        println("[PDV] Cannot plot '$path': $(sprint(showerror, err))")
    end
    nothing
end

function pdv_handle(obj::AbstractVector{<:Real}, path::String, tree::PDVTree)
    _plot_with_makie(path) do mk
        Base.invokelatest(getproperty(mk, :lines), obj;
                          axis=(title=path,))
    end
    nothing
end

function pdv_handle(obj::AbstractMatrix{<:Real}, path::String, tree::PDVTree)
    _plot_with_makie(path) do mk
        fig, ax, hm = Base.invokelatest(getproperty(mk, :heatmap), obj;
                                        axis=(title=path,))
        Base.invokelatest(getproperty(mk, :Colorbar), fig[1, 2], hm)
        fig
    end
    nothing
end

# 0-D and ≥3-D numeric arrays: a friendly notice with `dispatched: true`,
# matching pdv-python's "[PDV] Cannot plot N-D ndarray" (second review —
# without this the dispatch fell through to `dispatched: false` + error,
# a kernel-visible response divergence between the backends).
function pdv_handle(obj::AbstractArray{<:Real}, path::String, tree::PDVTree)
    println("[PDV] Cannot plot $(ndims(obj))-D ndarray (size=$(size(obj))); " *
            "default handler supports 1D and 2D only.")
    nothing
end

# ---------------------------------------------------------------------------
# Complex-array defaults (parity with pdv-python's complex plotting)
# ---------------------------------------------------------------------------

# 1-D complex: one axis with labeled Re and Im lines.
function pdv_handle(obj::AbstractVector{<:Complex}, path::String, tree::PDVTree)
    _plot_with_makie(path) do mk
        fig = Base.invokelatest(getproperty(mk, :Figure))
        ax = Base.invokelatest(getproperty(mk, :Axis), fig[1, 1]; title=path)
        Base.invokelatest(getproperty(mk, :lines!), ax, real.(obj); label="Re")
        Base.invokelatest(getproperty(mk, :lines!), ax, imag.(obj); label="Im")
        Base.invokelatest(getproperty(mk, :axislegend), ax)
        fig
    end
    nothing
end

# 2-D complex: side-by-side Re/Im heatmaps, each with its own colorbar
# (the ranges are generally unrelated, so a shared bar would mislead).
function pdv_handle(obj::AbstractMatrix{<:Complex}, path::String, tree::PDVTree)
    _plot_with_makie(path) do mk
        fig = Base.invokelatest(getproperty(mk, :Figure))
        axr = Base.invokelatest(getproperty(mk, :Axis), fig[1, 1]; title="Re")
        hmr = Base.invokelatest(getproperty(mk, :heatmap!), axr, real.(obj))
        Base.invokelatest(getproperty(mk, :Colorbar), fig[1, 2], hmr)
        axi = Base.invokelatest(getproperty(mk, :Axis), fig[1, 3]; title="Im")
        hmi = Base.invokelatest(getproperty(mk, :heatmap!), axi, imag.(obj))
        Base.invokelatest(getproperty(mk, :Colorbar), fig[1, 4], hmi)
        Base.invokelatest(getproperty(mk, :Label), fig[0, :], path)
        fig
    end
    nothing
end

function pdv_handle(obj::AbstractArray{<:Complex}, path::String, tree::PDVTree)
    println("[PDV] Cannot plot $(ndims(obj))-D complex array (size=$(size(obj))); " *
            "default handler supports 1D and 2D only.")
    nothing
end
